core/src/main/cljc/jarl/builtins/http.cljc

```diff
@@ -191,9 +208,54 @@
                   "raw_body"    (:body response)}]
         resp))))
 
+(defn- date [res]
+  (when-let [date (get-in res [:headers "date"])]
+    (time/http-date->instant date)))
+
+(defn- max-age [res]
+  (when-let [cc (get-in res [:headers "cache-control"])]
+    (first (->> (str/split (str/lower-case cc) #",")
+                (remove str/blank?)
+                (map str/trim)
+                (map #(str/split % #"="))
+                (filter #(= "max-age" (first %)))
+                (map second)
+                (map parse-long)))))
+
+(defn- exp-from-headers [res]
+  (if-let [max-age (max-age res)]
+    (some-> (date res) (t/>> (t/new-duration max-age :seconds)))
+    (-> res (get-in [:headers "expires"]) time/http-date->instant)))
+
+(defn- expires-at [{now :time-now-ns} req res]
+  (if (force-cache? req)
+    (t/>> (time/ns->instant now)
+          (t/new-duration (:force-cache-duration-seconds req) :seconds))
+    (or (exp-from-headers res)
+        (time/ns->instant now))))
+
+(defn- maybe-cache [bctx res req]
+  (let [res (cond-> res (:headers res) (update :headers update-keys str/lower-case))]
+    (when (cache? req)
+      (:response (swap! inter-query-cache assoc (hash req) {:response res :exp (expires-at bctx req res)})))
+    res))
+
+; side effect - may remove stale item from cache
+(defn- check-cache [{now :time-now-ns} key]
+  (when (contains? @inter-query-cache key)
+    (let [item (get @inter-query-cache key)
+          exp  (:exp item)]
+      (if (t/>= (time/ns->instant now) exp)
+        (do ; cache item is stale — remove and return nil to fetch anew
+          (swap! inter-query-cache dissoc key)
```

Comment:
See the docs on [atoms](https://clojure.org/reference/atoms):
> As with all reference types, the intended use of atom is to hold one of Clojure’s immutable data structures. And, similar to ref’s alter and agent’s send, you change the value by applying a function to the old value. This is done in an atomic manner by [swap!](https://clojure.github.io/clojure/clojure.core-api.html#clojure.core/swap!) Internally, [swap!](https://clojure.github.io/clojure/clojure.core-api.html#clojure.core/swap!) reads the current value, applies the function to it, and attempts to compare-and-set! it in. Since another thread may have changed the value in the intervening time, it may have to retry, and does so in a spin loop. The net effect is that the value will always be the result of the application of the supplied function to a current value, atomically. However, because the function might be called multiple times, it must be free of side effects.